File: go/weightedgraph/index_min_heap.go
package weightedgraph

import (
	"errors"
)

type IndexMinHeap struct {
	data []*Edge
	indexes []int
	reverse []int
	count int
	capacity int
}

func NewIndexMinHeap(cap int) *IndexMinHeap {
	heap := &IndexMinHeap{}
	heap.data = make([]*Edge, cap)
	heap.indexes = make([]int, cap)
	heap.reverse = make([]int, cap)

	// -1 代表元素还不在堆中
	for i := 0; i < cap; i++ {
		heap.reverse[i] = -1
	}
	heap.capacity = cap
	return heap
}

func (this *IndexMinHeap) GetSize() int {
	return this.count
}

func (this *IndexMinHeap) IsEmpty() bool {
	return this.count == 0
}

func (this *IndexMinHeap) Insert(index int, value *Edge) error {
	// 满容量或者索引超出
	if this.count == this.capacity || index >= this.capacity || index < 0 {
		return errors.New("Error: No availablle space or index out of range")
	}

	// 在插入元素前，需要判断索引 i 不在堆中，否则就堆中就会有两个索引 i 指向了同
	// 一元素，在添加的时候不会出问题，但是在出堆的时候会出问题
	if this.contains(index) {
		return errors.New("Element had been in heap")
	}

	this.data[index] = value
	// 一定要注意这里，indexes 中只有我们添加进的元素
	// 也就是并不是所有的数据都在堆中有代表的索引
	this.indexes[this.count] = index
	this.reverse[index] = this.count
	this.count++
	this.shiftUp(this.count)
	return nil
}

func (this *IndexMinHeap) contains(index int) bool {
	if index < 0 || index >= this.capacity {
		return false
	}
	return this.reverse[index] != -1
}

func (this *IndexMinHeap) ExtractMin() (*Edge, error) {
	if this.count == 0 {
		return nil, errors.New("Empty Heap")
	}
	value := this.data[this.indexes[0]]
	this.swapIndex(0, this.count-1)
	this.reverse[this.indexes[this.count-1]] = -1
	this.count--
	this.shiftDown(0)
	return value, nil
}

func (this *IndexMinHeap) ExtractMinIndex() (int, error) {
	if this.count == 0 {
		return 0, errors.New("Empty Heap")
	}
	maxIndex := this.indexes[0]
	this.swapIndex(0, this.count-1)
	this.reverse[this.indexes[this.count-1]] = -1
	this.count--
	this.shiftDown(0)
	return maxIndex, nil
}

func (this *IndexMinHeap) GetMin() (*Edge, error) {
	if this.count == 0 {
		return nil, errors.New("Empty Heap")
	}
	value := this.data[this.indexes[0]]
	return value, nil
}

func (this *IndexMinHeap) GetMinIndex() (int, error) {
	if this.count == 0 {
		return 0, errors.New("Empty Heap")
	}
	value := this.indexes[0]
	return value, nil
}

func (this *IndexMinHeap) GetItem(index int) (*Edge, error) {
	if !this.contains(index) {
		return nil, errors.New("Invalid index")
	}
	return this.data[this.indexes[index]], nil
}

func (this *IndexMinHeap) Change(index int, value *Edge) error {
	if !this.contains(index) {
		return errors.New("Invalid index")
	}

	this.data[index] = value
	// for i := 0; i < this.count; i++ {
	// 	if this.indexes[i] == index {
	// 		this.shiftUp(i)
	// 		this.shiftDown(i)
	// 		return nil
	// 	}
	// }

	j := this.reverse[index]
	this.shiftUp(j)
	this.shiftDown(j)
	return nil
}

func (this *IndexMinHeap) parent(i int) int {
	return (i-1) / 2
}

func (this *IndexMinHeap) left(i int) int {
	return i * 2 + 1
}

func (this *IndexMinHeap) right(i int) int {
	return i * 2 + 2
}

func (this *IndexMinHeap) shiftUp(index int) {
	for ; index > 0; {
		pIndex := this.parent(index) 
		// 比较是 indexes 中指向的数据的值
		if this.data[this.indexes[index]].Weight() < this.data[this.indexes[pIndex]].Weight() {
			this.swapIndex(index, pIndex)
			index = pIndex
		} else {
			break
		}
	}
}

func (this *IndexMinHeap) shiftDown(index int) {
	for ; this.left(index) < this.count; {
		l := this.left(index)
		j := l
		if l+1 < this.count && this.data[this.indexes[l+1]].Weight() < this.data[this.indexes[l]].Weight() {
			j = l+1
		}

		if this.data[this.indexes[j]].Weight() < this.data[this.indexes[index]].Weight() {
			this.swapIndex(j, index)
			index = j
		} else {
			break
		}
	}
}

func (this *IndexMinHeap) swapIndex(i, j int) {
	temp := this.indexes[i]
	this.indexes[i] = this.indexes[j]
	this.indexes[j] = temp

	this.reverse[this.indexes[i]] = i
	this.reverse[this.indexes[j]] = j
}

File: js/graphSP/indexMinHeap.js
class IndexMinHeap {
  constructor(cap) {
    this.cap = cap;
    this.data = Array(cap);
    this.indexes = Array(cap);
    this.reverse = Array(cap);
    this.count = 0;

    for (let i = 0; i < cap; i++) {
      this.reverse[i] = -1;
    }
  }

  IsEmpty() {
    return this.count === 0;
  }

  GetSize() {
    return this.count;
  }

  Contains(index) {
    if (index < 0 || index >= this.cap) {
      return false;
    }
    return this.reverse[index] !== -1; 
  }

  Insert(index, value) {
    if (index < 0 || index >= this.cap) {
      throw new Error("Invalid index");
    }

    if (this.Contains(index)) {
      throw new Error("Index exists");
    }

    this.data[index] = value;
    this.indexes[this.count] = index;
    this.reverse[index] = this.count;
    this.shiftUp(this.count);
    this.count++;
  }

  GetMin() {
    if (this.count === 0) {
      throw new Error("Empry heap");
    }
    return this.data[this.indexes[0]];
  }

  GetMinIndex() {
    if (this.count === 0) {
      throw new Error("Empry heap");
    }
    return this.indexes[0];
  }

  GetItem(index) {
    if (!this.Contains(index)) {
      return null;
    }
    return this.data[index];
  }

  ExtractMin() {
    if (this.count === 0) {
      throw new Error("Empry heap");
    }

    const value = this.data[this.indexes[0]];
    this.reverse[this.indexes[0]] = -1;
    this.swapIndex(0, this.count-1);
    this.count--;
    this.shiftDown(0);
    return value;
  }

  ExtractMinIndex() {
    if (this.count === 0) {
      throw new Error("Empry heap");
    }

    const index = this.indexes[0];
    this.reverse[index] = -1;
    this.swapIndex(0, this.count-1);
    this.count--;
    this.shiftDown(0);
    return index;
  }

  Change(index, value) {
    if (!this.Contains(index)) {
      throw new Error("Invalid index");
    }

    this.data[index] = value;
    const j = this.reverse[index];
    this.shiftUp(j);
    this.shiftDown(j);
  }

  parent(index) {
    return Math.floor((index - 1) / 2);
  }

  leftChild(index) {
    return index * 2 + 1;
  }

  rightChild(index) {
    return index * 2 + 1;
  }

  shiftUp(index) {
    while(index > 0 && this.data[this.indexes[index]] < this.data[this.indexes[this.parent(index)]]) {
      this.swapIndex(index, this.parent(index));
      index = this.parent(index);
    }
  }

  shiftDown(index) {
    while(this.leftChild(index) < this.count) {
      let j = this.leftChild(index);
      if (j+1 < this.count && this.data[this.indexes[j+1]] < this.data[this.indexes[j]]) {
        j++;
      }

      if (this.data[this.indexes[j]] < this.data[this.indexes[index]]) {
        this.swapIndex(j, index);
        index = j;
      } else {
        break;
      }
    }
  }

  swapIndex(i, j) {
    const temp = this.indexes[j];
    this.indexes[j] = this.indexes[i];
    this.indexes[i] = temp;
    this.reverse[this.indexes[i]] = i;
    this.reverse[this.indexes[j]] = j;
  }
}

module.exports = IndexMinHeap;
